src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/DefaultConverter.cs

```diff
@@ -0,0 +1,64 @@
+        {
+            // Filter out identifiers which are not used in statements.
+            var symbolNames = new HashSet<string>(ForEachInfo.Statements.SelectMany(statement => semanticModel.AnalyzeDataFlow(statement).ReadInside).Select(symbol => symbol.Name));
+            var identifiersUsedInStatements = ForEachInfo.Identifiers.Where(identifier => symbolNames.Contains(identifier.ValueText));
+
+            // If there is a single statement and it is a block, leave it as is.
+            // Otherwise, wrap with a block.
+            var block = WrapWithBlockIfNecessary(ForEachInfo.Statements.Select(statement => statement.KeepCommentsAndAddElasticMarkers()));
+
+            editor.ReplaceNode(ForEachInfo.ForEachStatement, CreateDefaultReplacementStatement(ForEachInfo, identifiersUsedInStatements, block).WithAdditionalAnnotations(Formatter.Annotation));
+        }
+
+        private StatementSyntax CreateDefaultReplacementStatement(
+            ForEachInfo<ForEachStatementSyntax, StatementSyntax> forEachInfo,
+            IEnumerable<SyntaxToken> identifiers,
+            BlockSyntax block)
+        {
+            var identifiersCount = identifiers.Count();
+            if (identifiersCount == 0)
+            {
+                // Generate foreach(var _ ... select new {})
+                return SyntaxFactory.ForEachStatement(VarNameIdentifier, SyntaxFactory.Identifier("_"), CreateQueryExpression(SyntaxFactory.AnonymousObjectCreationExpression(), Enumerable.Empty<SyntaxToken>(), Enumerable.Empty<SyntaxToken>()), block);
+            }
+            else if (identifiersCount == 1)
+            {
+                // Generate foreach(var singleIdentifier from ... select singleIdentifier)
+                return SyntaxFactory.ForEachStatement(VarNameIdentifier, identifiers.Single(), CreateQueryExpression(SyntaxFactory.IdentifierName(identifiers.Single()), Enumerable.Empty<SyntaxToken>(), Enumerable.Empty<SyntaxToken>()), block);
+            }
+            else
+            {
+                var tupleForSelectExpression = SyntaxFactory.TupleExpression(SyntaxFactory.SeparatedList(identifiers.Select(identifier => SyntaxFactory.Argument(SyntaxFactory.IdentifierName(identifier)))));
+                var declaration = SyntaxFactory.DeclarationExpression(
+                    VarNameIdentifier,
+                    SyntaxFactory.ParenthesizedVariableDesignation(
+                        SyntaxFactory.SeparatedList<VariableDesignationSyntax>(identifiers.Select(identifier => SyntaxFactory.SingleVariableDesignation(identifier)))));
+
+                // Generate foreach(var (a,b) ... select (a, b))
+                return SyntaxFactory.ForEachVariableStatement(declaration, CreateQueryExpression(tupleForSelectExpression, Enumerable.Empty<SyntaxToken>(), Enumerable.Empty<SyntaxToken>()), block);
+            }
+        }
+
+        private static BlockSyntax WrapWithBlockIfNecessary(IEnumerable<StatementSyntax> statements)
```

Comment:
this seems can be generic helper as well.

Reply:
I think it is too small method to extract. The best place I found was SyntaxFactory but the method seems to be too specific for it.

Similar (not same) methods are declared in CSharpMethodExtractor.CSharpCodeGenerator

---
In reply to: [195849628](https://github.com/dotnet/roslyn/pull/27130#discussion_r195849628) [](ancestors = 195849628)